src/openforms/forms/tasks.py

```diff
@@ -96,6 +96,29 @@ def repopulate_reusable_definition_variables_to_form_variables(form_id: int) ->
         FormVariable.objects.create_for_form(form)
 
 
+@app.task(ignore_result=True)
+@transaction.atomic()
+def create_form_variables_for_components(form_id: int) -> None:
+    """Create FormVariables for each component in the Form
+
+    This task is scheduled after creating/updating a FormStep, to ensure that the saved
+    Form has the appropriate FormVariables, even if errors occur in the variables update
+    endpoint. This is done to avoid leaving the Form in a broken state.
+
+    See also: https://github.com/open-formulieren/open-forms/issues/4824#issuecomment-2514913073
+    """
+    from .models import Form, FormVariable  # due to circular import
+
+    form = Form.objects.get(id=form_id)
+
+    # delete the existing form variables, we will re-create them
+    FormVariable.objects.filter(
+        form=form_id, source=FormVariableSources.component
+    ).delete()
+
+    FormVariable.objects.create_for_form(form)
```

Comment:
shouldn't this also be followed by the recouple task? E.g. when you're updating a form that has existing submissions.

Reply:
Hmm yeah that's right and I think the `repopulate_reusable_definition_variables_task` needs to be run as well, because if the FormDefinition that you're updating is reusable, those other Forms need their FormVariables updated as well